### ISSUE TITLE
Error: no workspace

### ISSUE THREAD

**Issue by farahfa** [jenkinsci/docker-plugin#325]
Hello,

I'm having an issue where once a build is done (fails or succeeds), then I cannot view the workspace anymore (because the container gets removed). So I get the following error:

![screen shot 2015-09-18 at 3 28 50 pm](https://cloud.githubusercontent.com/assets/9218600/9969585/5f4423cc-5e1d-11e5-9963-33a12aa4766a.png)

I've mounted a volume to the container hoping that I can still view the workspace even when the container is destroyed, but that doesn't work either. Here is my container template:

![screen shot 2015-09-18 at 3 25 28 pm](https://cloud.githubusercontent.com/assets/9218600/9968948/a1c5e16c-5e19-11e5-81cc-a75d1034ce42.png)

Thanks in advance,

Fadi

**Comment by farahfa** [comment]
Just to clarify, the error shows up only when the job is done. While it's running I can view the workspace with no problems.

**Comment by KostyaSha** [comment]
When you are running it WS mapping is not used and ws is shown directly.
Fro WS mapping your jenkins_home should be rw mounted to docker host and then binded with volume.

**Comment by farahfa** [comment]
Thanks for the fast reply! Ah, so I'm configuring it incorrectly! I'm sorry for labeling this as an issue, I thought I did everything possible.

Could you please clarify how to what you told me? I'm still new to this plugin, so consider me as a beginner.

**Comment by farahfa** [comment]
@KostyaSha, could you please elaborate on what you said? Because I believe I've done what you've said but I keep on having the same issue.

**Comment by tomaxsas** [comment]
You have to mount your workspace folder from docker host to Jenkins master server, and then in Remote FS root write where you mounted it, without workspace at the end.

**Comment by KostyaSha** [comment]
Usually people just mount workspace folder from jenkins master, or jenkins_home into containers via NFS.

**Comment by farahfa** [comment]
Ah okay, Thanks @tomaxsas & @KostyaSha, just as a suggestion (don't know if it's feasible or not, but), would it be easier to add a feature to the plugin later that can point the workspace directly to the docker host? So for example: one would attach a volume (so the WS can be stored outside the container) and then from Jenkins master global config, one would add the ip address and location of where the WS is and that would read them?

Anyway, I appreciate your help very much! Thank you.

**Comment by KostyaSha** [comment]
There is a problem with it, people can share folder using different tools/FSs. Atm afaik all cloud slaves just sharing JENKINS_HOME. Ideally only folder from exact build path on jenkins master should be rw mounted to the slave. But job may also have different workspaces and tools. If you will have some good design/architecture proposal, then of course it would make sense add it.

**Comment by kummerer94** [comment]
Hey everyone,

I like this plugin very much, it takes away so much work and the newest versions provision and tear down slaves very smoothly.

I want to add a use-case to this issue. However, I don't know whether this is actually possible using Jenkins:

If I want to make the workspace of a job in Jenkins available to the slaves, I have to mount the whole workspace (in my case that's a lot of workspaces and I don't want jobs or slaves to have access to other job's workspaces).

We want to have a similar functionality like the CloudBees Docker Custom Build Environment provides. But, since the plugin poorly handles the startup of the job (not everything is run inside of the container), we can't and won't use it. (Also, you can't configure every option like `--add-host`, etc.) We need to start a slave for a specifiy job and after that job is finished the slave is torn down.

Thus, the slave's lifecycle is bound to the job's. My idea basically would be to provide another availability option that allows us to configure a once used slave that gets deleted after the job is finished.

The goal of this is to be able to use the Jenkins Job environment (at least the core variables, like `JOB_NAME`) during startup. This would allow us to only mount job specific directories and solve the issue of mounting the whole workspace.

Furthermore, you could pull specific images using tags based on these variables.

Hope to hear your feedback,
Alex

**Comment by KostyaSha** [comment]
@kummerer94 this issue is about viewing workspace internals after build end. As i understand you want other thing.

**Comment by kummerer94** [comment]
Yes, but it seemed heavily related. Should I create a new issue?

**Comment by KostyaSha** [comment]
If you want see workspace always - you should mount folder into slave run. There is nothing here for fixing. If you want something else, then fill new issue.

**Comment by thomassuckow** [comment]
I believe cloudbees has a plugin that will copy the workspace back to master. I don't use cloudbees though so I cannot confirm.

**Comment by KostyaSha** [comment]
I suggest discuss it as separate issue :( Existed functionality works as expected. For new features i have ideas but will be glad to see use cases in separate threads. TOO hard find something later.

**Comment by kummerer94** [comment]
Thanks for the suggestion, @thomassuckow. But as I described above, the Cloudbees plugin is not really ready for production and many things are still processed on the Jenkins master and not in the docker container.

I will open another issue and describe my use case there. Thanks for the response!

**Comment by David-Villeneuve** [comment]
After a lot of pain trying to get this to work, I finally figured out why this was not working for me. 

I had the NFS storage mapped to /build_data on my Jenkins master, so all the workspaces were  under /build_data/workspace.  The "Remote FS Root Mapping" was set to "/build_data".  This did not work.

It turns out that I was using a custom workspace definition in my jobs, such that the last portion of the path was not the same as the job name.   I think the plug-in is looking for "/build_data/workspace/<job-name>", irrespective of the last portion of the path for the custom workspace definition of the job. 

Creating a symbolic link to the workspace corresponding to the job name (and restarting Jenkins to pick it up) solved the problem (of course, making the workspace name match the job does too).

**Comment by jagadeeswar725863** [comment]
@David-Villeneuve. It helps a lot for me.. But New version of docker-plugin. i couldn't find "Remote FS Root Mapping" option.. so i am still getting "Error: No workspace" Could you pls suggest... Below is the architecture for my jenkins / slave.

--> Two servers, one for jenkins master and another one for docker host to run agent as a container. we are using docker-plugin to run agent as a container. We have nfs share as a workspace which is mounted in Docker host, container and Jenkins master. 

Thanks in advance......

**Comment by David-Villeneuve** [comment]
@jagadeeswar725863 I did see that this feature was deprecated, and I see that it's gone now.  In the end, I did not use the feature given that I needed to run concurrent jobs/containers, meaning I needed a dynamic workspace per container.  It got worse, as I also needed to load-share across more than one docker host.  I tried my best to have some kind of persistent yet dynamically allocated storage per container, but in the end my scheme for this became to complex, and there was no way to close all the critical sections.  

However, for a single non-concurrent slave/container, I can see support for viewing a persistent workspace after container slave completes being useful for debugging, but it breaks the way that viewing the workspace is normally done, which is through the slave agent (which breaks here because the agent is non-persistent). 

For debugging jobs I take two approaches.

Firstly, I try to put any meaningful output into log files, and then make sure that I capture these as a build artifact.  This will use storage on the Jenkins server (vs slave) when it completes, so you can't keep too much, so you have to be very thoughtful about this.

Secondly, if I really need to look into the container, I'll add a very long sleep as a final job step, and then go onto the docker host and use the docker exec command (docker exec -i -t <container-id> /bin/bash) to get into a shell on the still-running-container and look around.  Also, until the container terminates, you can still view the workspace.
